### DISCUSSION TITLE
How to structure multiple sequences of graphs

### DISCUSSION THREAD

**Discussion by petergroth** [pyg-team/pytorch_geometric#3048] [Q&A] [answered]
Hi,

I have a dataset which consists of sequences of graphs which describe the state of different systems evolving over time. Let's assume that the edge structure is static and that the number of nodes for each sequence is constant, and that only the node features change from time step to time step. What would be a suitable way of handling the Data() objects and creating a Pytorch Geometric Dataset?

A simple approach would be to create a Data() object for each graph for each time step, and then setting shuffle to False when creating the dataloader. This might lead to batching issues, however, at it seems inflexible. Alternatively, since I'm (currently at least) assuming a static edge structure, I can create a single Data() object for each sequence, and then simply have the time step as an additional dimension of the node attributes.

What is the suggested approach to handling multiple sequences of graphs while retaining the flexibility of the Dataset and DataLoader classes?

**Comment by rusty1s** [comment, marked as answer]
If your edge structure is static and only the node features evolve over time, you can save your node features as `[num_nodes, num_timestamps, num_features]` inside `Data` (as you suggested). This should support mini-batching out-of-the box. For edge structures, the current `DataLoader` sadly cannot handle sequences, which I want to fix in an upcoming release. This should allow you to do the following:
```python
data.edge_index = [edge_index_1, edge_index_2, ...]
```
There also exists [PyTorch Geometric Temporal](https://github.com/benedekrozemberczki/pytorch_geometric_temporal) which might already fit your use-case.

**Reply by MCiurletti** [reply]
Hi,

may I ask how sequence support is currently progressing? I have a similar dataset as explained in the question, but with dynamic edge structure and I'm struggling to find a good solution. I'm currently trying to make it work with PyG Temporal, however all their temporal datasets consist of a single dynamic graph and I have a hard time adapting that to my needs without being unreasonably inefficient.

**Reply by rusty1s** [reply]
PyG 2.0.* does support this already :)

**Reply by MCiurletti** [reply]
Oh wow, I guess I didn't look close enough. Thanks!

**Reply by MCiurletti** [reply]
Just one more question, as you recommended to use `[num_nodes, num_timestamps, num_features]` node feature shape, how could you handle additional or disappearing nodes? Same goes for edge features. Intuitively I would say that `num_timesteps` makes more sense as first dimension.

**Reply by rusty1s** [reply]
Yes, it makes more sense as first dimension, but shouldn't make much of a difference.

Adding or removing nodes is still challenging, and I suggest for a given mini-batch, you simply use dummy nodes (zero features, no graph connectivity) before a certain node appears. Does that work in your case?

**Reply by MCiurletti** [reply]
Okay, something like that was my idea too. Thanks for the help!
